src/HealthChecks/HealthChecks/test/HealthCheckPublisherHostedServiceTest.cs

```diff
@@ -385,6 +486,103 @@ public async Task RunAsync_CanFilterHealthChecks()
+
+        var service = CreateService(
+            publishers,
+            configurePublisherOptions: (options) =>
+            {
+                options.Predicate = (r) => r.Name.Contains("Delay");
+            },
+            configureBuilder: b =>
+            {
+                b.Add(
+                    new(
+                        name: "CheckDefault",
+                        instance: new DelegateHealthCheck(_ => Task.FromResult(HealthCheckResult.Healthy(HealthyMessage))),
+                        failureStatus: null,
+                        tags: null));
+
+                b.Add(
+                    new HealthCheckRegistration(
+                        name: "CheckDelay2Period18",
+                        instance: new DelegateHealthCheck(_ => Task.FromResult(HealthCheckResult.Healthy(HealthyMessage))),
+                        failureStatus: null,
+                        tags: null)
+                    {
+                        Delay = TimeSpan.FromSeconds(2),
+                        Period = TimeSpan.FromSeconds(18)
+                    });
+
+                b.Add(
+                    new HealthCheckRegistration(
+                        name: "CheckDelay7Period11",
+                        instance: new DelegateHealthCheck(_ => Task.FromResult(HealthCheckResult.Healthy(HealthyMessage))),
+                        failureStatus: null,
+                        tags: null,
+                        timeout: default)
+                    {
+                        Delay = TimeSpan.FromSeconds(7),
+                        Period = TimeSpan.FromSeconds(11)
+                    });
+
+                b.Add(
+                   new HealthCheckRegistration(
+                       name: "CheckDelay9Period5",
+                       instance: new DelegateHealthCheck(_ =>
+                       {
+                           unblockDelayedCheck.TrySetResult(null); // Unblock last delayed check
+                           return Task.FromResult(HealthCheckResult.Healthy(HealthyMessage));
+                       }),
+                       failureStatus: null,
+                       tags: null,
+                       timeout: default)
+                   {
+                       Delay = TimeSpan.FromSeconds(9),
+                       Period = TimeSpan.FromSeconds(5)
+                   });
+            });
+
+        try
+        {
+            await service.StartAsync();
+
+            // Act
+            await RunServiceAsync(service).TimeoutAfter(TimeSpan.FromSeconds(20));
+
+            await unblockDelayedCheck.Task;
+
+            // Assert
+            for (var i = 0; i < publishers.Length; i++)
+            {
+                var entries = publishers[i].Entries.SelectMany(e => e.report.Entries.Select(e2 => e2.Key)).OrderBy(k => k).ToArray();
+
+                Assert.True(entries.Count() == 3);
+                Assert.Equal(
+                    new[] { "CheckDelay2Period18", "CheckDelay7Period11", "CheckDelay9Period5" },
```

Comment:
We should also shorten the delays by a factor of 10 in this test and check the entries after the finally when we know the timers are stopped.